README
*  $Id: README,v 1.3 2001/08/14 11:07:42 davee Exp $
*
*  DosUnix - Convert DOS text line breaks to Unix text line breaks
*  Copyright (C)1997-2000 Dave Ewart (davee@sungate.co.uk)
*
*  This program is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program; if not, write to the Free Software
*  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 
NOTES

DosUnix is actually a pair of programs: dosunix and unixdos, which
perform conversion of line breaks between DOS text files and Unix text
files.  Note that files given as arguments are assumed to be text files.

ChkTxt is a third program supplied with the DosUnix suite, which is used
to identify which text format a given file is in.


NEWS
Version 1.0.11, 14.08.2001

First public release.

Version 1.0.12, November 2001

Actually, _this_ is the first public release.

Version 1.0.13, April 2002

Minor code cleanup

Version 1.0.14, September 2005


const.h
// $Id: const.h,v 1.13 2005/09/09 12:14:57 daveewart Exp $

/*
 * 
 * DosUnix - File converter - changes DOS end-of-lines to Unix
 * end-of-lines
 * Copyright (C)1997-2005 Dave Ewart (davee@sungate.co.uk)
 *
 * This file is part of the DosUnix project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define REVDATE "09.09.2005"
#define AUTHOR "Dave Ewart"
#define AUTHOREMAIL "davee@sungate.co.uk"
#define COPYYEAR "1997-2005"



common.c
// $Id: common.c,v 1.12 2005/09/09 12:04:48 daveewart Exp $

#include <stdio.h>
#include <stdlib.h>
#include "const.h"

/*
 * 
 * DosUnix - File converter - changes DOS end-of-lines to Unix
 * end-of-lines
 * Copyright (C)1997-2002 Dave Ewart (davee@sungate.co.uk)
 *
 * This file is part of the DosUnix project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

int NoInput(char *somefile);
int NoOutput(char *somefile);
int PrintSummary(char *progname, long NumChanges);
int BasicInfo(char *progname);
int UsageInfo(char *command, int style);
int ShowOptions(char *command);

int NoInput(char *somefile)
{
    printf("ERROR: Cannot open input file: %s\n", somefile);
    return (0);
}

int NoOutput(char *somefile)
{
    printf("ERROR: Cannot open output file: %s\n",somefile);
    return (0);
}

int PrintSummary(char *progname, long NumChanges)
{
    printf("%s %s - SUMMARY: A total of %ld ", progname, VERSION,
           NumChanges);
    printf("end of line characters were modified.\n");
    return (0);
}

int BasicInfo(char *progname)
{
    printf("%s version %s ", progname, VERSION);
    printf("(C)%s %s (%s)\n", COPYYEAR, AUTHOR, AUTHOREMAIL);
    printf("%s is part of the DosUnix project.\n",progname);
    return (0);
}

int FullInfo(char *progname)
{
    printf("Last revised %s\n\n", REVDATE);
    printf
        ("%s comes with ABSOLUTELY NO WARRANTY.  This is free software\n",
         progname);
    printf
        ("and you are welcome to redistribute it under certain conditions.\n");
    printf("Read the file COPYING supplied with this distribution, or\n");
    printf("http://www.gnu.org/licenses/gpl.txt for details.\n\n");
    return (0);
}

int UsageInfo(char *command, int style)
{
    printf("Usage:\n");
    if (style == 1) {
        printf("  %s INPUTFILE\n\n", command);
    }
    else {
        printf("  %s INPUTFILE OUTPUTFILE\n\n", command);
    }
    return (0);
}

int ShowOptions(char *command)
{
    printf("  %s -V  for version and license info;\n", command);
    printf("  %s -h  for usage info\n", command);
    return (0);
}


dosunix.c
// $Id: dosunix.c,v 1.12 2005/09/09 12:04:48 daveewart Exp $

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * DosUnix - File converter - changes DOS end-of-lines to Unix
 * end-of-lines
 * Copyright (C)1997-2002 Dave Ewart (davee@sungate.co.uk)
 *
 * This file is part of the DosUnix project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

extern int NoInput(char *somefile);
extern int NoOutput(char *somefile);
extern int PrintSummary(char *progrname, long NumChanges);
extern int BasicInfo(char *progname);
extern int FullInfo(char *progname);
extern int UsageInfo(char *command, int style);
extern int ShowOptions(char *command);
extern int main(int argc, char *argv[])
{
    FILE *InputFile, *OutputFile;
    int c;
    long NumChanges = 0;

    if (argc == 1) {
        c = BasicInfo("DosUnix");
        c = ShowOptions("dosunix");
        exit(0);
    }
    if (argc == 2) {
        if ((strcmp(argv[1], "-V")) == 0) {
            c = BasicInfo("DosUnix");
            c = FullInfo("DosUnix");
        }
        else if ((strcmp(argv[1], "-h")) == 0) {
            c = BasicInfo("DosUnix");
            c = UsageInfo("dosunix", 2);
            printf("DosUnix converts files from DOS text ");
            printf("format to Unix text format\n");
            printf("by replacing each carriage return & ");
            printf("newline pair with a single\n");
            printf("newline character.\n\n");
        }
        else {
            printf("Unknown option: %s\n", argv[1]);
        }
        exit(0);
    }

    if (!(InputFile = fopen(argv[1], "r"))) {
        c = NoInput(argv[1]);
        c = UsageInfo("dosunix", 2);
        exit(0);
    };
    if (!(OutputFile = fopen(argv[2], "w"))) {
        c = NoOutput(argv[2]);
        c = UsageInfo("dosunix", 2);
        exit(0);
    };
    while (1) {
        c = getc(InputFile);
        if (c == -1) {
            break;
        }
        else if (c == 13) {
            if ((c = getc(InputFile)) == 10) {
                putc(10, OutputFile);
                NumChanges++;
            }
        }
        else {
            putc(c, OutputFile);
        }
    }
    fclose(InputFile);
    fclose(OutputFile);
    c = PrintSummary("DosUnix", NumChanges);
    return (0);
}


unixdos.c
// $Id: unixdos.c,v 1.12 2005/09/09 12:04:48 daveewart Exp $

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * 
 * UnixDos  - File converter - changes Unix end-of-lines to DOS
 * end-of-lines
 * Copyright (C)1997-2002 Dave Ewart (davee@sungate.co.uk)
 *
 * This file is part of the DosUnix project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

extern int NoInput(char *somefile);
extern int NoOutput(char *somefile);
extern int PrintSummary(char *progname, long NumChanges);
extern int BasicInfo(char *progname);
extern int FullInfo(char *progname);
extern int UsageInfo(char *command, int style);
extern int ShowOptions(char *command);

int main(int argc, char *argv[])
{
    FILE *InputFile, *OutputFile;
    int c;
    long NumChanges = 0;

    if (argc == 1) {
        c = BasicInfo("UnixDos");
        c = ShowOptions("unixdos");
        exit(0);
    }
    if (argc == 2) {
        if ((strcmp(argv[1], "-V")) == 0) {
            c = BasicInfo("UnixDos");
            c = FullInfo("UnixDos");
        }
        else if ((strcmp(argv[1], "-h")) == 0) {
            c = BasicInfo("UnixDos");
            c = UsageInfo("unixdos", 2);
            printf("UnixDos converts files from Unix text ");
            printf("format to DOS text format\n");
            printf("by replacing each newline character with ");
            printf("a carriage return and\nline-feed pair.\n\n");
        }
        else {
            printf("Unknown option: %s\n", argv[1]);
        }
        exit(0);
    }

    if (!(InputFile = fopen(argv[1], "r"))) {
        c = NoInput(argv[1]);
        c = UsageInfo("unixdos", 2);
        exit(0);
    };
    if (!(OutputFile = fopen(argv[2], "w"))) {
        c = NoOutput(argv[2]);
        c = UsageInfo("unixdos", 2);
        exit(0);
    };
    while (1) {
        c = getc(InputFile);
        if (c == -1) {
            break;
        }
        else if (c == 10) {
            putc(13, OutputFile);
            putc(10, OutputFile);
            NumChanges++;
        }
        else {
            putc(c, OutputFile);
        }
    }
    fclose(InputFile);
    fclose(OutputFile);
    c = PrintSummary("UnixDos", NumChanges);
    return (0);
}


chktxt.c
// $Id: chktxt.c,v 1.9 2005/09/09 12:04:48 daveewart Exp $

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
 * 
 * ChkTxt
 * Determines, by inspection, whether a given text file is in DOS or
 * Unix text format.
 * Copyright (C)1997-2002 Dave Ewart (davee@sungate.co.uk)
 *
 * This file is part of the DosUnix project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

extern int NoInput(char *somefile);
extern int PrintSummary(char *progrname, long NumChanges);
extern int BasicInfo(char *progname);
extern int FullInfo(char *progname);
extern int UsageInfo(char *command, int style);
extern int ShowOptions(char *command);
extern int main(int argc, char *argv[])
{
    FILE *InputFile;
    int c;
    long n_dos   = 0;
    long n_unix  = 0;
    long n_dummy = 0;
    char cmd[256];

    if (argc == 1) {
        c = BasicInfo("ChkTxt");
        c = ShowOptions("chktxt");
        exit(0);
    }
    if (argc == 2) {
        if ((strcmp(argv[1], "-V")) == 0) {
            c = BasicInfo("ChkTxt");
            c = FullInfo("ChkTxt");
            exit(0);
        }
        else if ((strcmp(argv[1], "-h")) == 0) {
            c = BasicInfo("ChkTxt");
            c = UsageInfo("chktxt", 1);
            printf("ChkTxt will attempt to determine, ");
            printf("by inspection, whether\n");
            printf("the given text file is in DOS text ");
            printf("format or Unix text format.\n");
            printf("If this fails to determine the file format, ");
            printf("the 'file' command\n");
            printf("is used to identify the file type.\n\n");
            exit(0);
        }
    }

    if (!(InputFile = fopen(argv[1], "r"))) {
        c = NoInput(argv[1]);
        c = UsageInfo("chktxt", 1);
        exit(0);
    };
    while (1) {
        c = getc(InputFile);
        if (c == -1) {
            break;
        }
        else if (c == 13) {
            if ((c = getc(InputFile)) == 10) {
                n_dos++;
            }
            else {
                n_dummy++;
            }
        }
        else if (c == 10) {
            n_unix++;
        }
    }
    fclose(InputFile);
    printf("ChkTxt %s: ", VERSION);
    /* Some DOS line-endings and no Unix line-endings suggests
     * file is DOS
     */
    if (n_dos > 0 && n_unix == 0) {
        printf("%s is a DOS text file.\n", argv[1]);
    }
    /* Some Unix line-endings and DOS line-endings suggests
     * file is Unix
     */
    else if (n_unix > 0 && n_dos == 0 && n_dummy == 0) {
        printf("%s is a Unix text file.\n", argv[1]);
    }
    /* This is very simplistic but will probably work OK
     * for most systems
     */
    else {
        printf("Cannot determine filetype of %s\n", argv[1]);
        printf("It is probably not a simple text file.\n\n");
        printf("Running 'file %s' ...\n\n", argv[1]);
        strcpy(cmd, "file ");
        strcat(cmd, argv[1]);
        system(cmd);
    }
    return (0);
}
